### ISSUE TITLE
Nullness Checker: Any instance type of a class must be a subtype of its declared type

### ISSUE THREAD

**Issue by t-rasmud** [typetools/checker-framework#2247]
If the declaration type of a class is `@A`, then every instance type of this class must be a subtype of `@A` and the checker-framework must report an error otherwise. Currently, the checker-framework does not perform this check.

Consider this example:

```
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

@NonNull class DeclaredClass{}

class ValidUseType {
    void test(){
        @Nullable DeclaredClass object;
    }
}
```
Since `DeclaredClass` has the type `@NonNull`, it should be illegal to write `@Nullable DeclaredClass object;`

Running the command `javac -processor nullness ValidUseType.java ` does not issue any warning.

**Comment by wmdietl** [comment]
Related discussion: #1751
The current implementation uses class declaration annotations only as a default for uses of that type, not as a restriction on all possible uses.
For example:

```java
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

@NonNull
class DeclaredClass{}

class validUseType {
    void test(){
         DeclaredClass object = null;
    }
}
```

Gives an incompatible assignment error.

As outlined in the design document, I agree that changing class declaration annotations to be upper bounds would be better.

**Comment by t-rasmud** [comment]
Fixed in 937286d2e404db5b49177c2ba6dd1d511aedb711

**Comment by wmdietl** [comment]
Didn't that commit only add the possibility to declare the default for type declarations?
The issue with e.g. the Nullness Checker doesn't seem fixed.

**Comment by mernst** [comment]
Werner is right: the original issue is not fixed.  I have added a test case at `checker/tests/nullness/Issue2247.java`.

**Comment by smillst** [comment]
This is already implemented in the framework.  The Nullness Checker changes this behavior by overriding isValidUse in NullnessVisitor.
